### PR TITLE
NSFS | Versioning | Documentation

### PR DESCRIPTION
**Pull request by romayalon**
Signed-off-by: Romy <romy2232@gmail.com>

### Explain the changes
1. 

### Issues: Fixed #xxx / Gap #xxx
1. 

### Testing Instructions:
1. 

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: docs/design/NsfsVersioning.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+* PUT, POST and COPY operations will create a new version of the object identified by a unique version id.
+* DELETE latest version will create a delete marker which is a dummy version identified by a  unique version id as well.
+* DELETE version id will delete the version completely.
+
+
+### Version ID 
+* Will be allocated to a version on PUT/POST request (or DELETE version id). 
+* Consisted of the version mtimeMS + inode number.
+* The version ID will be attached as an xattr of the file.
+* When latest versions is moved to past versions directory, the file name will be changed from key to key_{version_id}.
```
Comment (liranmauda) [inline docs/design/NsfsVersioning.md:47]:
```suggestion
* The version ID will be attached as a xattr of the file.
* When the latest version is moved to the past versions directory, the file name will be changed from key to key_{version_id}.
```

File: docs/design/NsfsVersioning.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+* A dummy file that will be created under the hidden .versions/ directory on DELETE latest request.
+* When A delete marker is the latest version of an object, it indicates that the object is deleted.
+* A unique version id will be allocated to the delete marker similiarly tor regular versions.
+  
+### Posix safe rename
+
+#### In the following cases NooBaa will move files between a directory and its .versions/ directory:
+
+* Put object / Complete multipart upload - a new latest version will be created and the old latest should move to .versions/.
+* Delete latest version - The latest version should move from parent directory to .versions/ & a delete marker will be created in .versions as well.
```
Comment (liranmauda) [inline docs/design/NsfsVersioning.md:60]:
```suggestion
* Delete latest version - The latest version should move from the parent directory to .versions/ & a delete marker will be created in .versions as well.
```

File: docs/design/NsfsVersioning.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+* Delete version id (latest version & regular version) - The current latest version should be removed and the second latest should move from .versions/ to the parent directory.
+* Delete version id (latest version & a delete marker) - The current latest version should be removed and if the second latest is not a delete marker it should move from .versions/ to the parent directory.
+
+In order to support best effort on scale of these scenarions, for POSIX file systems, we will use the following methods -
```
Comment (liranmauda) [inline docs/design/NsfsVersioning.md:64]:
```suggestion
In order to support best effort on scale of these scenarios, for POSIX file systems, we will use the following methods - 
```

**Review by liranmauda — APPROVED** [review]
Reading this is clear to me, added some minor suggestions.

File: docs/design/NsfsVersioning.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+* A unique version id will be allocated to the delete marker similiarly tor regular versions.
+  
+### Posix safe rename
+
+#### In the following cases NooBaa will move files between a directory and its .versions/ directory:
+
+* Put object / Complete multipart upload - a new latest version will be created and the old latest should move to .versions/.
+* Delete latest version - The latest version should move from parent directory to .versions/ & a delete marker will be created in .versions as well.
+* Delete version id (latest version & regular version) - The current latest version should be removed and the second latest should move from .versions/ to the parent directory.
+* Delete version id (latest version & a delete marker) - The current latest version should be removed and if the second latest is not a delete marker it should move from .versions/ to the parent directory.
```
Comment (shirady) [inline docs/design/NsfsVersioning.md:62]: I think those cases are harder to understand, maybe an illustrative diagram (or a link) will be helpful.
Does the last case ('Delete version id (latest version & a delete marker)') is recursive? If yes, maybe it is worth mentioning it.

File: docs/design/NsfsVersioning.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+### Delete marker
+* A dummy file that will be created under the hidden .versions/ directory on DELETE latest request.
+* When A delete marker is the latest version of an object, it indicates that the object is deleted.
+* A unique version id will be allocated to the delete marker similiarly tor regular versions.
+  
+### Posix safe rename
+
+#### In the following cases NooBaa will move files between a directory and its .versions/ directory:
+
+* Put object / Complete multipart upload - a new latest version will be created and the old latest should move to .versions/.
```
Comment (shirady) [inline docs/design/NsfsVersioning.md:59]:
```suggestion
* Put object / Complete multipart upload - a new latest version will be created and the "old latest" should move to .versions/.
```

File: docs/design/NsfsVersioning.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+* Will be allocated to a version on PUT/POST request (or DELETE version id). 
+* Consisted of the version mtimeMS + inode number.
+* The version ID will be attached as an xattr of the file.
+* When latest versions is moved to past versions directory, the file name will be changed from key to key_{version_id}.
+
+
+### Delete marker
+* A dummy file that will be created under the hidden .versions/ directory on DELETE latest request.
+* When A delete marker is the latest version of an object, it indicates that the object is deleted.
+* A unique version id will be allocated to the delete marker similiarly tor regular versions.
```
[the remark this answers is not in the document]
Comment (romayalon) [reply docs/design/NsfsVersioning.md]: I changed the 'regular' usage of directories to 'original'

File: docs/design/NsfsVersioning.md (hunk 4 above)
Comment (romayalon) [reply docs/design/NsfsVersioning.md:62]: I added a link to S3 versioning doc, in there, you will find that behind-the-scenes versions of the same object are stacked. I think that this idea is important in order to understand these scenarios.
Also, illustrations can be found in the detailed downstream documentation.
The last case is just a deletion of a version that is the latest version but it's also a delete marker

File: docs/design/NsfsVersioning.md (hunk 1 above)
Comment (romayalon) [reply docs/design/NsfsVersioning.md]: https://www.pristineword.com/grammar-an-x/#:~:text=It's%20an%20X.,the%20indefinite%20article%20%22an%22.

**Pull request merged**